wolfcrypt/src/asn.c

```diff
@@ -19936,6 +19936,118 @@ int wc_PemPubKeyToDer(const char* fileName,
 
 #endif /* !NO_FILESYSTEM && WOLFSSL_PEM_TO_DER */
 
+/* Get public key in DER format from ASN.1 encoded X.509 certificate.
+ *
+ * x509   - input certificate in DER/ASN.1 format
+ * x509Sz - size of x509, octets
+ * type   - type of x509, CERT_TYPE for X.509 cert, or CERTREQ_TYPE for CSR
+ * format - format of x509, CTC_FILETYPE_PEM (PEM) or CTC_FILETYPE_ASN1 (DER).
+ *          CTC_FILETYPE_PEM is currently not supported with
+ *          WOLFSSL_NO_MALLOC.
+ * derKey - output buffer to place DER/ASN.1 encoded public key
+ * derKeySz [IN/OUT] - size of derKey buffer on input, size of public key
+ *                     on return. If derKey is passed in as NULL, derKeySz
+ *                     will be set to required buffer size for public key
+ *                     and LENGTH_ONLY_E will be returned from function.
+ * Returns 0 on success, or negative error code on failure. LENGTH_ONLY_E
+ * if derKey is NULL and returning length only.
+ */
+int wc_GetPubKeyDerFromX509(const byte* x509, word32 x509Sz, int type,
+                            int format, byte* derKey, word32* derKeySz)
+{
+    int ret;
+#ifdef WOLFSSL_SMALL_STACK
+    DecodedCert* decoded;
+#else
+    DecodedCert decoded[1];
+#endif
+#ifndef WOLFSSL_NO_MALLOC
+    DerBuffer* der = NULL;
+#endif
+    byte* certDer = (byte*)x509;
+    word32 certDerSz = x509Sz;
+
+    /* derKey may be NULL to return length only */
+    if (x509 == NULL || x509Sz == 0 || derKeySz == NULL ||
+        (derKey != NULL && *derKeySz == 0) ||
+        (type != CERT_TYPE && type != CERTREQ_TYPE) ||
+        (format != CTC_FILETYPE_PEM && format != CTC_FILETYPE_ASN1)) {
+        return BAD_FUNC_ARG;
+    }
+#ifndef WOLFSSL_CERT_REQ
+    if (type == CERTREQ_TYPE) {
+        WOLFSSL_MSG("For CSR support, wolfSSL must be compiled "
+                    "with --enable-certreq");
+        return BAD_FUNC_ARG;
+    }
+#endif
+#ifdef WOLFSSL_NO_MALLOC
+    if (format == CTC_FILETYPE_PEM) {
+        WOLFSSL_MSG("PEM not supported with WOLFSSL_NO_MALLOC");
+        return BAD_FUNC_ARG;
+    }
+#endif
+
+#ifdef WOLFSSL_SMALL_STACK
+    decoded = (DecodedCert*)XMALLOC(sizeof(DecodedCert), NULL,
+                                    DYNAMIC_TYPE_TMP_BUFFER);
+    if (decoded == NULL) {
+        return MEMORY_E;
+    }
+#endif
+
+#ifndef WOLFSSL_NO_MALLOC
```

Comment:
Is there a reason to not also get on the `WOLFSSL_PEM_TO_DER` macro?

Reply:
Agreed, should be guarded by this as well.  May remove this code, pending discussion below.